site/src/main/paradox/Scio-Unit-Tests.md

```diff
@@ -82,3 +82,43 @@ To run the test, we use the `runWithContext`, this will run calculateTeamScores
 @@snip [LeaderBoardTest.scala](/scio-examples/src/test/scala/com/spotify/scio/examples/complete/game/LeaderBoardTest.scala) { #LeaderBoardTest_example_3 }
 
 Scio provides more `SCollection` assertions such as `inWindow`, `inCombinedNonLatePanes`, `inFinalPane`, and `inOnlyPane`. You can find the full list [here](https://spotify.github.io/scio/api/com/spotify/scio/testing/SCollectionMatchers.html). More information on testing unbounded pipelines can be found [here](https://beam.apache.org/blog/2016/10/20/test-stream.html).
+
+### Test with transform overrides
+
+Scio provides a method to replace arbitrary _named_ `PTransform`s in a test context; this is primarily useful for mocking requests to external services.
+
+In this example, the `GuavaLookupDoFn` stands in for a transform that contacts an external service.
+A `ParDo` `PTransform` is created from the `DoFn` (`ParDo.of`), then applied to the pipeline (`applyTransform`) with a unique name (`myTransform`).
+
+@@snip [JobTestTest.scala](scio-test/src/test/scala/com/spotify/scio/testing/JobTestTest.scala) { #JobTestTest_example_1 }
+
+In a `JobTest`, a `PTransformOverride` can be passed to the `transformOverride` method to replace transforms in the original pipeline.
+Scio provides convenience methods for constructing `PTransformOverride`s in the `com.spotify.scio.testing.TransformOverride` object.
+Continuing the example above, `TransformOverride.ofAsyncLookup` can be used to map static mock data into the expected output format
+for the transform, here `KV[Int, BaseAsyncLookupDoFn.Try[String]]`.
+
+@@snip [JobTestTest.scala](scio-test/src/test/scala/com/spotify/scio/testing/JobTestTest.scala) { #JobTestTest_example_2 }
+
+It is also possible to provide a function rather than a static map:
+
+@@snip [JobTestTest.scala](scio-test/src/test/scala/com/spotify/scio/testing/JobTestTest.scala) { #JobTestTest_example_3 }
+
+`TransformOverride.of` overrides transforms of type `PTransform[PCollection[T], PCollection[U]]` as in the case of `BaseAsyncDoFn` subclasses.
+`TransformOverride.ofKV` overrides transforms of type `PTransform[PCollection[T], PCollection[KV[T, U]]]`.
+
+Sources can also be overridden with `TransformOverride.ofSource`. For example, this source:
+
+@@snip [JobTestTest.scala](scio-test/src/test/scala/com/spotify/scio/testing/JobTestTest.scala) { #JobTestTest_example_4 }
+
+Can be overridden with static mock data:
+
+@@snip [JobTestTest.scala](scio-test/src/test/scala/com/spotify/scio/testing/JobTestTest.scala) { #JobTestTest_example_5 }
+
+Due to type erasure it is possible to provide the incorrect types for the transform and the error will not be caught until runtime.
+Typically a `ClassCastException` will be thrown whose message contains the correct type:
```

Comment:
The error message is rather unhelpful. Would it be possible to catch that specific error and provide a more informative message ?

Reply:
Sort of. Basically all we have control over is inside the PTransform, so if we construct it such that it knows what type the user asserts it should have in tests, and we witness an element input during the test that isn't assignable, then we can preemptively throw a nice exception. I've added this.

It however doesn't help for anything with type parameters like `KV` which will still 💥 with `ClassCastException`.

Also, we don't have any information about the true output type, either in the test or in the PTransform, so if we return the wrong output type we'll always get a `ClassCastException` from a coder.